### PR TITLE
Develop 509 runfolder organisation service

### PR DESCRIPTION
**Pull request by b97pla**
**What problems does this PR solve?**

This PR introduces an OrganisationService for organizing a runfolder in preparation for staging and delivery. Upon organisation, project data (fastq files and sequence reports) will be symlinked to a project-specific directory in the runfolder. The hierarchy will be  `Projects/<project name>/<runfolder name>/`. Each sample will be placed in its own subdirectory. A project-specific file with checksums will be placed in the organized folder, as well as a sample sheet where all entries not belonging to the project has been masked.

In summary:

* An API endpoint and corresponding OrganisationHandler are introduced.
* the Runfolder and Project models have been extended
* models for Sample and SampleFile objects have been added
* Repositories for managing Samples, Unorganized projects and runfolders have been added
* services have been extended
* the test suite has been expanded with an integration test, unit tests and utility functions

**How has the changes been tested?**

the test suite includes unit tests and an integration test. In addition, I've run the service in a Docker container and tested organisation with a range of input alternatives. 

**Reasons for careful code review**

This code will have a direct effect on what data is made available to the downstream delivery operations and as such, should be carefully code reviewed and tested. The code will not remove any data per se but silently failing to organize or mixing up samples would have serious consequences for what is delivered to the end user.

  - [X]

### CODE REVIEW

File: delivery/repositories/runfolder_repository.py [hunk 1]
```diff
@@ -107,3 +128,116 @@ def get_project(self, project_name):
         for project in self.get_projects():
             if project.name == project_name:
                 yield project
+
+    def dump_project_checksums(self, project):
+        raise NotImplementedError()
```
[the remark this answers is not in the document]
Comment (johandahlberg) [reply delivery/repositories/runfolder_repository.py]: Just looking around in the code it seems to me that this is being used, so should it in fact be implemented?

File: delivery/repositories/runfolder_repository.py [hunk 2]
```diff
@@ -107,3 +128,116 @@ def get_project(self, project_name):
...
+                * lane in project samples
+            """
+            return self.project_repository.is_sample_in_project(
+                project,
+                e.get("Sample_Project"),
+                e.get("Sample_ID"),
+                # e.g. MiSeq SampleSheets may not have the Lane column, so assume 1 if missing
+                int(e.get("Lane", "1")))
+
+        def _mask_samplesheet_entry(e):
```
Comment (johandahlberg) [inline delivery/repositories/runfolder_repository.py]: This is a smart idea.

File: delivery/repositories/runfolder_repository.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (b97pla) [reply delivery/repositories/runfolder_repository.py]: it's only implemented in the subclass and it doesn't really fill any purpose to have it as not implemented in the parent class, so I can certainly remove it here

**Comment by b97pla** [comment]
Thanks for the input, I'd be happy to do a walkthrough

File: delivery/repositories/runfolder_repository.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (johandahlberg) [reply delivery/repositories/runfolder_repository.py]: No. That sounds fine in that case. However, if there is only one subclass of the parent class (as far as I can see from a quick look at the PR), is it really necessary to use inheritance here at all?

**Comment by b97pla** [comment]
this should be good to merge now!

**Pull request merged**